### PR TITLE
Chordal sparsity

### PR DESCRIPTION
**Pull request by blegat**
Rebasing of https://github.com/JuliaOpt/SumOfSquares.jl/pull/128 on top of master branch

- [x] Create a `SparsePutinar` certificate
- [x] Create tests from examples

### CODE REVIEW

File: src/csp/ChordalExtensionGraph.jl [hunk 1]
```diff
@@ -170,32 +170,22 @@ function is_clique(G::Graph{T}, x::Vector{T}) where T
...
     # (less fill-in, earlier elimination)
-    σ = sortperm(fill_in.(H.graph, 1:num_nodes))
-    elimination_order = Dict(H.int2n[σ[i]] => i for i in 1:num_nodes)
+    σ = sortperm(fill_in.(H, 1:num_nodes))
+    elimination_order = zeros(Int, num_nodes)
+    for i in 1:num_nodes
+        elimination_order[σ[i]] = i
+    end
 
     # Computes elimination graph `H` using equation (6.1) of [VA15].
```
Comment (tweisser) [inline src/csp/ChordalExtensionGraph.jl]: What is [VA15] ?

**Comment by tweisser** [comment]
_greedy_triangulation! should reset the disabled field before returning candidate_cliques:
using SumOfSquares.ChordalExtensionGraph
G = LabelledGraph{Int}()
add_node!.(G, [1,2,3,4,5])
add_edge!(G, 1, 2)
add_edge!(G, 2, 3)
add_edge!(G, 3, 4)
add_edge!(G, 1, 4)

G,_ = chordal_extension(G, ChordalExtensionGraph.GreedyFillIn())

add_edge!(G, 1, 5)

G,_ = chordal_extension(G, ChordalExtensionGraph.GreedyFillIn())

"""
ERROR: Maximal cliques do not cover all nodes.
Stacktrace:
 [1] error(::String) at ./error.jl:33
 [2] chordal_extension(::SumOfSquares.ChordalExtensionGraph.Graph, ::SumOfSquares.ChordalExtensionGraph.GreedyFillIn) at /home/tweisser/.julia/dev/SumOfSquares/src/csp/ChordalExtensionGraph.jl:277
 [3] chordal_extension(::LabelledGraph{Int64}, ::SumOfSquares.ChordalExtensionGraph.GreedyFillIn) at /home/tweisser/.julia/dev/SumOfSquares/src/csp/ChordalExtensionGraph.jl:296
 [4] top-level scope at none:0
"""

**Comment by blegat** [comment]
> _greedy_triangulation! should reset the disabled field before returning candidate_cliques

Done in 659ff0a9440d82ea8d0b4985dd0e47a47153c6a3

**Comment by tweisser** [comment]
In view of https://arxiv.org/abs/1912.08899 I would replace sparse::Bool by a keyword sparse::SparsityFlag defaulting to NoSparsity.

abstract type SparsityFlag end
struct NoSparsity <: SparsityFlag
struct ChordalSparsity  <: SparsityFlag
struct TermSparsity  <: SparsityFlag

**Comment by blegat** [comment]
https://arxiv.org/abs/1912.08899 seems to employ two ways to distinguish between the two sparsity approaches:
1) Monomial sparsity vs Variables sparsity where the type of the nodes in the chordal graphs are either monomials or variables.
2) Term sparsity vs correlative sparsity where the name correlative sparsity seems to come from "Sums of squares and semidefinite program relaxations for polynomial optimization problems with structured sparsity"

I prefer the namings form 1) since the namings from 2) seems the name correlative seems to have been chosen before there was a need to distinguish between the two so it's not clear that this naming makes a clear distinction.

**Pull request merged**